File: Simple Updater Library/SimpleUpdater.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace Simple_Updater_Library
{
    public partial class SimpleUpdater
    {
        #region Public vars Declaration

        private string server_url;
        public string Server_Url
        {
            get => this.server_url;
        }

        private string installation_path;
        public string Installation_path
        {
            get => this.installation_path;
        }

        private bool canDownload;
        public bool CanDownload
        {
            get => canDownload;
        }

        private bool canCheck;
        public bool CanCheck
        {
            get => canCheck;
        }

        private long numberOfBytesToDownload;
        public long NumberOfBytesToDownload
        {
            get => this.numberOfBytesToDownload;
        }

        private uint statusCode;
        public uint StatusCode
        {
            get => this.statusCode;
        }

        #endregion

        #region Events Declaration

        public delegate void CheckFinished(int nbrLocalFiles, int nbrServerFiles, int nbrFilesToDownload, int nbrFilesDeleted, long numberOfBytesToDownload);
        public event CheckFinished OnCheckFinished;

        public delegate void DownloadFinished(long totalbytesdownloaded);
        public event DownloadFinished OnDownloadFinished;

        public delegate void DownloadProgressChanged(long totalbytesdownloaded, float percent);
        public event DownloadProgressChanged OnDownloadProgressChanged;

        public delegate void StatusChanged(uint code);
        public event StatusChanged OnStatusChanged;

        #endregion

        #region Private Variables Declaration

        // Var - Global of the Updater (Check + Download)
        private Dictionary<string, File> server_files;
        private List<string> ignore_list_files;
        private Dictionary<string, bool> ignore_list_folders;

        #endregion

        #region Constructor

        public SimpleUpdater(string server_url, string installation_path)
        {
            this.server_url = server_url;
            this.installation_path = installation_path;
            canDownload = false;
            canCheck = true;
            statusCode = 0;
        }

        #endregion

        public void CheckFilesFromServerAndDeleteOutdated()
        {
            if(!canCheck)
            {
                throw new Exception("Check not authorized. Are you downloading ?");
            }

            server_files = new Dictionary<string, File>();

            this.canCheck = false;
            this.canDownload = false;

            Thread checker = new Thread(() => Check());
            checker.Start();
        }

        public void DownloadFiles()
        {
            if(!canDownload || server_files == null)
            {
                throw new Exception("Download not authorized. Have you checked out server files ?");
            }

            this.canDownload = false;
            this.canCheck = false;
            DownloadFiles_Downloader();
        }

        #region Events Redirection

        private void Check_Finished(int nbrLocalFiles, int nbrServerFiles, int nbrFilesToDownload, int nbrFilesDeleted)
        {
            this.canCheck = true;
            this.canDownload = true;

            this.OnCheckFinished(nbrLocalFiles, nbrServerFiles, nbrFilesToDownload, nbrFilesDeleted, this.numberOfBytesToDownload);
            Status_Changed(3);
            Status_Changed(0);
        }

        private void Download_Finished()
        {
            this.canDownload = true;
            this.canCheck = true;
            this.OnDownloadFinished(this.bytesdownloaded);
            Status_Changed(8);
            Status_Changed(0);
        }

        private void DownloadProgress_Changed(long totalbytesdownloaded)
        {
            float percent;
            // Division by 0 ? Don't know if it can occur
            try
            {
                percent = (float)(totalbytesdownloaded) / (float)numberOfBytesToDownload;
            }
            catch
            {
                percent = 100;
            }
            
            // Trigger event
            OnDownloadProgressChanged(totalbytesdownloaded, percent * 100);
        }

        private void Status_Changed(uint code)
        {
            /*
             * 0 - Waiting for action
             * 1 - Contacting server for files to parse
             * 2 - Search and delete local files
             * 3 - Check finished - Triggered after Check_Finished(...) - (not very usefull - duplicate of Check_Finished(...) )
             * 4 - Download started
             * 5 - A file has been successfully downloaded
             * 6 - An error occurered when downloading a file
             * 7 - Download cancelled
             * 8 - Download finished - Triggered after Download_Finished(...) - (not very usefull - duplicate of Download_Finished(...) )
             */

            this.statusCode = code;
            this.OnStatusChanged(code);
        }

        #endregion
    }
}

File: JAVA - Simple Updater Library/src/fr/s0me1ne/simpleupdater/SimpleUpdater.java
package fr.s0me1ne.simpleupdater;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SimpleUpdater {

	// Public vars declaration
	
	private String server_url;
    private String installation_path;
    private Boolean canDownload;
    private Boolean canCheck;
    private long numberOfBytesToDownload;
    private int statusCode;

    public String getServerUrl()
    {
    	return this.server_url;
    }
    
    public String getInstallationPath()
    {
    	return this.installation_path;
    }
    
    public Boolean CanDownload()
    {
    	return this.canDownload;
    }
    
    public Boolean CanCheck()
    {
    	return this.canCheck;
    }
    
    public long getNumberOfBytesToDownlaod()
    {
    	return this.numberOfBytesToDownload;
    }
    
    public int getStatusCode()
    {
    	return this.statusCode;
    }
	
    
    // Var - Global of the Updater (Check + Download)
    private Map<String, File> server_files;
    private List<SimpleUpdaterEvents> listeners = new ArrayList<SimpleUpdaterEvents>();
    Checker checker;
    Downloader downloader;
    GetEvent getEvents = new GetEvent(this);
	
    public SimpleUpdater(String server_url, String installation_path)
    {
        this.server_url = server_url;
        this.installation_path = installation_path;
        this.canDownload = false;
        this.canCheck = true;
        this.statusCode = 0;
        addListener(getEvents);
    }
    
    public void addListener(SimpleUpdaterEvents e)
    {
    	this.listeners.add(e);
    }
    
    public void CheckFilesFromServerAndDeleteOutdated()
    {
        if(!canCheck)
        {
            //throw new Exception("Check not authorized. Are you downloading ?");
        	System.out.println("Check not authorized. Are you downloading ?");
        	return;
        }

        server_files = new HashMap<String, File>();

        this.canCheck = false;
        this.canDownload = false;

        checker = new Checker(this.installation_path, this.server_url, this.listeners);
        Thread threadChecker = new Thread(checker);
        threadChecker.start();
    }
    
    public void DownloadFiles()
    {
        if(!canDownload || server_files == null)
        {
            //throw new Exception("Download not authorized. Have you checked out server files ?");
        	System.out.println("Download not authorized. Have you checked out server files ?");

        	return;
        }

        this.canDownload = false;
        this.canCheck = false;
        downloader = new Downloader(this.installation_path, this.server_url, checker.server_files, checker.numberOfBytesToDownload, this.listeners);
        downloader.DownloadFiles_Downloader();
    }
    
    void setCanCheckAndDownloadTrue()
    {
    	this.canCheck = true;
    	this.canDownload = true;
    }

}

class GetEvent implements SimpleUpdaterEvents
{
	SimpleUpdater updater;
	
	GetEvent(SimpleUpdater updater)
	{
		this.updater = updater;
	}

	@Override
	public void OnCheckFinished(int nbrLocalFiles, int nbrServerFiles, int nbrFilesToDownload, int nbrFilesDeleted,
			long numberOfBytesToDownload) {
		updater.setCanCheckAndDownloadTrue();		
	}

	@Override
	public void OnDownloadFinished(long totalbytesdownloaded) {
		updater.setCanCheckAndDownloadTrue();
	}

	@Override
	public void OnDownloadProgressChanged(long totalbytesdownloaded, float percent) {}

	@Override
	public void OnStatusChanged(int code) {}
	
}
